Init grist core at 1.5.1

Scandiravian opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This is based on the work done in ~~#305019 and~~ #322633. I've added some additional changes to the service and changed the `buildPhase` to bring the output size down with a few hundred MiB.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RaitoBezarius committed:
nixos/tests/grist-core: init

Basic smoketest for the gVisor sandboxing.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

bendlas committed:
pythonPackages.friendly-traceback: init

needed for grist-core

Co-authored-by: phaer <hello@phaer.org>

soyouzpanda commented:
![image](https://github.com/user-attachments/assets/c18cdb67-d625-48f5-bca8-88eeb90ad0e5)

Scandiravian commented:
> ![image](https://private-user-images.githubusercontent.com/23421201/406152343-c18cdb67-d625-48f5-bca8-88eeb90ad0e5.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2NTQ3NjAsIm5iZiI6MTczNzY1NDQ2MCwicGF0aCI6Ii8yMzQyMTIwMS80MDYxNTIzNDMtYzE4Y2RiNjctZDYyNS00OGY1LWJjYTgtODhlZWI5MGFkMGU1LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIzVDE3NDc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmZjI3ZDcwMDY5ZDdjOGNhY2QxOGM5MTc3YjhiNzMzODljYTIxMzBlZDI0ZWEwMDBhOTc1NWI3OWJiMjhmNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YOjZb84wPCXTqTxQnum2oMVtH8SLHy_1jfl5YUlnwyw)

I'm not sure I follow - could you explain what you mean by this? :confused:

Scandiravian commented:
> ![image](https://private-user-images.githubusercontent.com/23421201/406152343-c18cdb67-d625-48f5-bca8-88eeb90ad0e5.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2NTQ3NjAsIm5iZiI6MTczNzY1NDQ2MCwicGF0aCI6Ii8yMzQyMTIwMS80MDYxNTIzNDMtYzE4Y2RiNjctZDYyNS00OGY1LWJjYTgtODhlZWI5MGFkMGU1LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIzVDE3NDc0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmZjI3ZDcwMDY5ZDdjOGNhY2QxOGM5MTc3YjhiNzMzODljYTIxMzBlZDI0ZWEwMDBhOTc1NWI3OWJiMjhmNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YOjZb84wPCXTqTxQnum2oMVtH8SLHy_1jfl5YUlnwyw)

@soyouzpanda I'm not sure I follow - could you explain what you mean by this? :confused:

Scandiravian commented:
> I do not want my work to be used in an open source project that collaborates with weapon makers and fascists, that's all.

I hear what you're saying. I don't want to get into a big discussion, but suffice to say that I understand your point of view and where you're coming from.

I can remove the commits that I cherry-picked from your PR and rewrite the module myself. I can't guarantee that it won't be somewhat similar to the work you made, since there's a limited number of ways to configure things in Nix, but it will remove your association from the history. Would that be acceptable to you?

Scandiravian commented:
> I do not want my work to be used in an open source project that collaborates with weapon makers and fascists, that's all.

@soyouzpanda I hear what you're saying. I don't want to get into a big discussion, but suffice to say that I understand your point of view and where you're coming from.

I can remove the commits that I cherry-picked from your PR and rewrite the module myself. I can't guarantee that it won't be somewhat similar to the work you made, since there's a limited number of ways to configure things in Nix, but it will remove your association from the history. Would that be acceptable to you?

bendlas commented:
Sorry, I don't really know what's going on and I also don't really feel like playing catch-up, so let me just write down what I'm taking from this and where I'm at:

I'm assuming that Scandiravian or their project has a known association with Anduril and the MIC, either way they don't seem to deny it. I am feeling blindsided by them not being up-front about it, because at this point, their controversial status within the community cannot be considered suprising. For this reason, I'll disengage from this conversation and rescind my earlier offer of helping shepherd their work. Please let me know if I've got anything wrong in my assessment.

thanks

Scandiravian commented:
> Sorry, I don't really know what's going on and I also don't really feel like playing catch-up, so let me just write down what I'm taking from this and where I'm at:
> 
> I'm assuming that Scandiravian or their project has a known association with Anduril and the MIC, either way they don't seem to deny it. I am feeling blindsided by them not being up-front about it, because at this point, their controversial status within the community cannot be considered suprising. For this reason, I'll disengage from this conversation and rescind my earlier offer of helping shepherd their work. Please let me know if I've got anything wrong in my assessment.
> 
> thanks

@bendlas I completely understand your position, but to clear things up, I don't have any affiliation with Anduril, nor any other company related to the MIC,  in any capacity. I never have and I never will, as it would be irreconcilable with my personal values.

The organisation I'm working for is a public institution that works to improve treatment for patients across the EU.

I think @soyouzpanda is uncomfortable contributing to nixpkgs as a whole, not due to anything related to me.

bendlas commented:
@Scandiravian thank you very much for that clarification! In this case, I'd like to ask your forgiveness for the misunderstanding and to re-offer my help. I'll have a closer look at this PR, next week.

bendlas commented:
As for soyouzpanda's contributions: It's probably best to remove their commits entirely, in order to respect their protest. 

I agree that it wouldn't be reasonable for them to expect zero overlap in solution space, and I feel if that was their goal they might have deleted their PR - lets just be as clean-room as possible, given that we've already looked at their commits.

cc @NixOS/moderation, just to make sure we're getting this right

Scandiravian commented:
> @Scandiravian thank you very much for that clarification! In this case, I'd like to ask your forgiveness for the misunderstanding and to re-offer my help. I'll have a closer look at this PR, next week.

There's nothing to forgive. This is a sensitive topic and I understand there are strong feelings involved.

Your help would be greatly appreciated!

I made some changes to the systemd unit that I forgot to push before finishing work yesterday. I got to a point where the module works with sandboxing disabled, but there's still some issues when it's turned on.  I'll push my work when I get to the office on Monday.

> As for soyouzpanda's contributions: It's probably best to remove their commits entirely, in order to respect their protest.
> 
> I agree that it wouldn't be reasonable for them to expect zero overlap in solution space, and I feel if that was their goal they might have deleted their PR - lets just be as clean-room as possible, given that we've already looked at their commits.
> 
> cc @NixOS/moderation, just to make sure we're getting this right

That sounds reasonable; until there's input from soyouzpanda on a solution that would work for them, I think it's the best we can do given the circumstances. I'll sort out the history on Monday.

Scandiravian committed:
grist-core: init at 1.5.1

Co-authored-by: phaer <hello@phaer.org>

Scandiravian committed:
nixos/grist-core: init

Co-authored-by: phaer <hello@phaer.org>

Scandiravian commented:
I've updated the history and pushed my local changes. The module should work as long as `enableSandboxing = false`. When it's enabled grist throws an error about `python3` not being available. I've checked that the binary is in the correct path in the systemd unit, so I'm not sure why the error happens.

It's something that could be fixed upstream by rewriting `sandbox/run.py` to not be so dependent on hardcoded paths, but I'm not sure how viable that is as it looks like a big undertaking.

I'll spend some more time on this issue later this week (probably Wednesday).

bendlas commented:
I gave this a quick whirl and got up to the failure with sandboxing, that you mentioned. I added two fixes, feel free to pull into your PR: https://github.com/bendlas/nixpkgs/compare/c21b41c634aff425f48d894012aaa51a32430c73...32ee8403354945f4580558ea3ef0fc29e5373c97

- the documentation build broke due to
  - the settings type being wrong
  - sandboxing violations (fixed with `defaultText`)
- when starting the first time with sandboxing, systemd could not mount the `grist-core/docs` folder
  - fixed by adding docs as a state directory

bendlas commented:
The sandboxing fail: `Exception: gvisor runsc problem: ["runsc", "-root", "/tmp/runsc", "-rootless", "-debug", "-network", "none", "run", "_tmp_tmpw0588zbf"]`, yes?

I didn't find anything so far, and I'll give this a break for now, but one useful command, that I used to get into the service context, maybe it can be helpful for you as well:

`nsenter -t $(systemctl show grist-core.service --property MainPID --value) -a --wd=/run/grist-core -e -S follow -- /nix/store/<hash>-bash-interactive-5.2p37/bin/bash --norc`

Erethon commented:
Thank you all for the work done already!

I'm testing this locally and as noted the nixos module works with sandboxing disabled. I also updated to the latest grist version (1.3.3) which includes your PR Scandiravian and tested without the patch and things seem to work. The diff is [here](https://gist.github.com/Erethon/fa022666177e28801e34058337696a0b).

I'll try looking at the sandboxing issue next.

Scandiravian commented:
> Thank you all for the work done already!
> 
> I'm testing this locally and as noted the nixos module works with sandboxing disabled. I also updated to the latest grist version (1.3.3) which includes your PR Scandiravian and tested without the patch and things seem to work. The diff is [here](https://gist.github.com/Erethon/fa022666177e28801e34058337696a0b).
> 
> I'll try looking at the sandboxing issue next.

That's really great; I appreciate your help!

I've done some more work on the PR this week and gotten a little bit further with the sandbox, though the progress is due to changes to `run.py` upstream, which I'm not a big fan of. 

The most important difference is that I mounted the nix store in the `runsc` sandbox, which fixed the `python not found` error, though `grist` now times out. I think this is due to the `Ready` message either not being send or read correctly.

I can share my changes when I'm back in the office Monday, though you might have gotten past where I am by then.

Once again, thanks for helping out with this ❤️

mweinelt commented on nixos/modules/services/web-apps/grist-core.nix:
```diff
@@ -0,0 +1,250 @@
+
+          EnvironmentFile = cfg.environmentFiles;
+        }
+        // (lib.optionalAttrs cfg.enableSandboxing {
```
It is surprising how this option mixes gvisor and systemd sandboxing. Why is systemd sandboxing made conditional anyway?

Scandiravian replied:
It was an attempt to circumvent the issue I've described here: https://github.com/NixOS/nixpkgs/pull/376176#issuecomment-2830312715

By sandboxing I could overwrite `/usr/bin/python3` with the python environment from the nix store

mweinelt commented on nixos/modules/services/web-apps/grist-core.nix:
```diff
@@ -0,0 +1,250 @@
+          ];
+        });
+
+      wantedBy = [ "multi-user.target" ];
```
Please move this up to after/wants, nobody will expect it down here.

mweinelt commented on nixos/tests/grist-core.nix:
```diff
@@ -0,0 +1,25 @@
+import ./make-test-python.nix (
```
Migrate this to `runtTest` instead, see https://github.com/NixOS/nixpkgs/issues/38687.

mweinelt commented on pkgs/by-name/gr/grist-core/package.nix:
```diff
@@ -0,0 +1,133 @@
+  src = fetchFromGitHub {
+    owner = "gristlabs";
+    repo = "grist-core";
+    rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```

mweinelt commented on pkgs/by-name/gr/grist-core/package.nix:
```diff
@@ -0,0 +1,133 @@
+    hash = "sha256-8/voGPKKlhTBAjBRLatg5Sf98hTSnKdice+Fhy73wTA=";
+  };
+
+  patches = [
+    # Upstream PR: https://github.com/gristlabs/grist-core/pull/1402
+    ./remove-chokidar.patch
+  ];
```
Merged in 1.3.3, please update and remove the vendored patch.

phaer commented:
@Scandiravian Are you still interested in this?

I rebased, updated to the latest version, applied most of the suggestions above (for grist-core, not friendly_traceback yet) and put in an ugly workaround for NoBrokenSymlinks. If that sounds interesting to you, please feel free to take from https://github.com/phaer/nixpkgs/tree/init-grist-core. Otherwise please let me know should you not plan to re-visit this :)

mweinelt commented on nixos/modules/services/web-apps/grist-core.nix:
```diff
@@ -0,0 +1,250 @@
+      description = "Group under which grist-core runs.";
+    };
+
+    pythonEnv = mkOption {
+      internal = true;
+      type = types.package;
+      default = pkgs.grist-core.pythonEnv;
+      example = literalExpression ''
+        pkgs.python3.withPackages (ps: with ps; [
+          astroid
+          asttokens
+          chardet
+          et-xmlfile
+          executing
+          friendly-traceback
+          iso8601
+          lazy-object-proxy
+          openpyxl
+          phonenumbers
+          pure-eval
+          python-dateutil
+          roman
+          six
+          sortedcontainers
+          stack-data
+          typing-extensions
+          unittest-xml-reporting
+          wrapt
+        ]);
+      '';
```
This is just a wrapper for `grist-core.pythonEnv`. If all we want is to create a name binding for that, then that should go into the let block.

Scandiravian commented:
> @Scandiravian Are you still interested in this?
> 
> I rebased, updated to the latest version, applied most of the suggestions above (for grist-core, not friendly_traceback yet) and put in an ugly workaround for NoBrokenSymlinks. It passes the vm test now, and the app starts but says it can't find python when tyring to create a new sheet. Probably the wrapper, but I am running out of time I am willing to invest for now. :/ EDIT: That was a missing sandbox variable while testing. Without gvisor one needs to set GRIST_SANDBOX to `lib.getExe pkgs.grist-core.pythonEnv`
> 
> If that sounds interesting to you, please feel free to take from https://github.com/phaer/nixpkgs/tree/init-grist-core. Otherwise please let me know should you not plan to re-visit this :)

Hey @phaer!

Thanks so much for working on this. 

I'm still planning to finish this up; things just got a bit busy, as they tend to do. I've written a to-do for myself to look at your changes tomorrow.

Scandiravian commented:
I've updated the PR with the changes from @phaer. Unfortunately there's still an issue when sandbox is enabled due to these lines in Grist: https://github.com/gristlabs/grist-core/blob/8d2ce5ed0d013ce1b08a90fa497a42f33cbd452a/sandbox/gvisor/run.py#L177-L191

This will result in the error `failed to load /usr/bin/python3: no such file or directory`, when using Grist through the web-interface

I'm not sure if there is a smart way to fake that path within the systemd unit

phaer commented:
> Unfortunately there's still an issue when sandbox is enabled due to these lines in Grist: https://github.com/gristlabs/grist-core/blob/8d2ce5ed0d013ce1b08a90fa497a42f33cbd452a/sandbox/gvisor/run.py#L177-L191
> 
> This will result in the error `failed to load /usr/bin/python3: no such file or directory`, when using Grist through the web-interface

To be clear, this is grist core only checking /usr/bin/python3 and /usr/local/bin/python3? Shouldn't this be handled by:

```
          BindReadOnlyPaths = [
            "${cfg.package.pythonEnv}/bin:/usr/bin"
            "${cfg.package.pythonEnv}/lib:/usr/lib"
          ];
```

?

Scandiravian commented:
> To be clear, this is grist core only checking /usr/bin/python3 and /usr/local/bin/python3? Shouldn't this be handled by:
> 
> ```
>           BindReadOnlyPaths = [
>             "${cfg.package.pythonEnv}/bin:/usr/bin"
>             "${cfg.package.pythonEnv}/lib:/usr/lib"
>           ];
> ```
> 
> ?

I think the error might be from `gvisor` and then propagated through grist-core. I'll look into it, when I'm back at the office tomorrow

Scandiravian commented:
I had a look at the issue today and yesterday

I changed the `run.py` script to mount the entire Nix store in the sandbox, which seems to have resolved the `failed to load /usr/bin/python3: no such file or directory` - the reason for this working is still unclear to me

Unfortunately the script now stalls at this call: https://github.com/gristlabs/grist-core/blob/51dc0d7949d483458c9eb8fd7b6d8feefa4aeaf6/sandbox/gvisor/run.py#L271. I'll try to look into this next issue tomorrow

My version of the script is here https://github.com/Scandiravian/grist-core/blob/nix-debug/sandbox/gvisor/run.py. It can be run by changing the source for `grist-core` to:

```nix
src = fetchFromGitHub {
  owner = "Scandiravian";
  repo = "grist-core";
  rev = "8023ac462fdf780055b198b7b2efa7f0797149f6";
  hash = "sha256-JAWPuVI3rf8EcRWND2JdeLjQnAV1mKEDU0HHB+JOFTg=";
};
```

My plan going forward is to fix the stalling issue, then figure out the root cause of both of them to determine whether it's possible to fix the issues in this PR or if it needs to be done upstream

mweinelt commented on nixos/tests/all-tests.nix:
Drop the formatting changes here.

phaer commented:
> My plan going forward is to fix the stalling issue, then figure out the root cause of both of them to determine whether it's possible to fix the issues in this PR or if it needs to be done upstream

Thank you for your investigations! I only had a quick glance at the source yet and might be misreading, but could it be that it failed to run python in the first place, because it needs to resolve a symlink pointing from the python binary to the nix store but the systemd sandbox didn't have that accessible?

It seems to resolve symlinks (realpath) while processing oci mount paths in the linked run.py, so a similar thing might happen one sandbox below, so to speak, for trying to run a binary in the nix store inside a "container"?

Scandiravian commented:
> Thank you for your investigations! I only had a quick glance at the source yet and might be misreading, but could it be that it failed to run python in the first place, because it needs to resolve a symlink pointing from the python binary to the nix store but the systemd sandbox didn't have that accessible?
> 
> It seems to resolve symlinks (realpath) while processing oci mount paths in the linked run.py, so a similar thing might happen one sandbox below, so to speak, for trying to run a binary in the nix store inside a "container"?

I got a bit further with the investigation. The gvisor python process starts as it should, but the interaction between the JS code and the python code which is somehow broken

Commands sent from the JS backend to the python process in the sandbox all time out. I'll try to figure out a way to debug this next week
